### PR TITLE
DD-296 : DDM, convert title to report

### PR DESCRIPTION
**Pull request by jo-pol**
Fixes DD-296 : DDM, convert title to report

When applied it will
--------------------
* fix the typo in `EasyConvertBagToDespositApp`
* if a title in the profile matches, a report number is added at the end of dcmiMetatdata
* if a title in dcmiMetatdata fully matches a report number, it is replaced
* if a title has a match followed with a "`:` and more text", a report number element is added
* about 18000 titles will become report numbers
* some 2600 titles will get an info log line: they might be report numbers
* another 200-300 "briefrapport" titles will be logged with rightsholder and/or publisher
  a test run will produce logging info for manual fixes or additional rules.
 
Where should the reviewer @DANS-KNAW/easy start @DANS-KNAW/dataversedans ?
------------------------------------------------

The core of the change is formed by the classes in the new `ddm` package.

Existing projects usually have one `RewriteRule` class with a multi-case pattern match for various transformations. I now chose for `RewriteRule` classes doing just one transformation. This reduces hard to read pattern matches to plain `if`s. It also provides more flexibility to compose `RuleTransformer` objects with different sets of `RewriteRule` objects.

~~The challenge for this pull request was the title element in the profile which should survive the transformation while similar titles in `dcmMetadata` should be replaced. Transformations by sub classes of `RewriteRule` are applied recursively and have no knowledge of siblings or parents. The challenge is solved using a `var` with a narrow scope in `DdmTransformer` which is _not_ a subclass of `RuleTransformer` though it has `transform` method with the same signature.~~

Another pull request will have to add provenance to the metadata of the bag. Preferably more fine grained than the current `EasyConvertBagToDepositApp.formatDiff`.  The `RewriteRule` objects are created at startup of the application. For a fine grained approach we need to somehow pass in rewrite loggers per dataset. To properly log the first title I'm thinking different `ReportRewriteRule` objects that only do the logging differently. See also my [comment](https://github.com/DANS-KNAW/easy-convert-bag-to-deposit/pull/11#discussion_r559445827) on another challenge for another pull request.

How should this be manually tested?
-----------------------------------

Related pull requests on github
-------------------------------

repo                       | PR                | note
-------------------------- | ----------------- | ----
easy-                      | [PR#](PRlink)     | ..

### CODE REVIEW

File: src/main/scala/nl/knaw/dans/easy/bag2deposit/ddm/DdmTransformer.scala [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+             { profileReportNumber }
+           </dcmiMetadata>.copy(prefix = n.prefix, attributes = n.attributes, scope = n.scope)
+    }
+  })
+
+  def transform(ddmIn: Node): Seq[Node] = {
+    val firstTitle = (ddmIn \ "profile" \ "title").flatMap(profileRuleTransformer)
+    profileReportNumber = firstTitle.filter(_.label == "reportNumber")
+    val ddmOut = ddmRuleTransformer(ddmIn)
+    logMissedBriefRapporten(firstTitle, ddmOut)
```
Comment (jo-pol) [inline src/main/scala/nl/knaw/dans/easy/bag2deposit/ddm/DdmTransformer.scala:50]: @lindareijnhoudt 
Would like to add the (not yet implemented) `LanguageRewriteRule` added to the `dcmiMetadataRuleTransformer`, but if not `(ddmIn \ "profile" \ "audience").text.contains("D37000"))` you want only the `LanguageRewriteRule`? Or do you only want to limit the title transformation to the specified audience?

**Review by janvanmansum — COMMENTED** [review]
* Don't throw `Exception`.
* Don't use `println` in tests (or basically anywhere else ;).
* Get CVS values by column name instead of column index.
* Test "it should "show matches and missed" In TitlesSpec is quite slow.

File: src/main/assembly/dist/cfg/ABR-reports.csv [hunk 2]
```diff
@@ -0,0 +1,135 @@
+cd0aec40-e199-43d7-927b-ee12f1fe7a31,"ADC-rapport","adc[^a-z]+rapport[a-z]*"
+dfd7f0b8-3423-410e-ad10-5f740906c1da,"ADT-rapport","adt[^a-z]+rapport[a-z]*"
+351707da-ab1d-438e-9d42-915d785594b3,"Aeres-Milieu Rapport","aeres[^a-z]+milieu[^a-z]+rapport[a-z]*"
+8e84ebc8-f379-4056-b171-8fbf54a76f15,"Aestimatica-rapport","aestimatica[^a-z]+rapport[a-z]*"
+6826d9aa-2b50-4f90-bc8c-902c56d75d22,"Alkmaar, Monumentenzorg en Archeologie Gemeente","alkmaar[^a-z]+monumentenzorg[^a-z]+en[^a-z]+ar(ch|g)a?eologi[a-z][^a-z]+gemeente"
+05c754af-7944-4971-8280-9e1b4e474a8d,"Rapporten over Alkmaarse Monumenten en Archeologie","(rapporten[^a-z]+over[^a-z]+alkmaar[a-z][^a-z]+monumenten[^a-z]+en[^a-z]+ar(ch|g)a?eologi[a-z]([^a-z]*rama)?|rama)?"
```
Comment (jo-pol) [inline src/main/assembly/dist/cfg/ABR-reports.csv:6]: regular expression should but does not find matches as shown by `matches-per-rce.txt` generated with [TitlesSpec](https://github.com/DANS-KNAW/easy-convert-bag-to-deposit/blob/d934f9e2a05b4bcfcb9343400a9e5280f042fda9/src/test/scala/nl.knaw.dans.easy.bag2deposit/TitlesSpec.scala#L64)

File: src/main/scala/nl/knaw/dans/easy/bag2deposit/ddm/ReportRewriteRule.scala [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+  private val trailer = "([.]|:.*)?"
+  val nrTailRegexp = s"$nrRegexp$trailer"
+  private val missedRegExp = s".*(notitie|rapport|bericht|publicat).*$nrRegexp$trailer"
+
+  val reportMap: Seq[ReportCfg] = parseCsv(cfgDir / "ABR-reports.csv", 0)
+    .map(r => ReportCfg(
+      uuid = r.get(0),
+      label = r.get(1),
+      regexp = r.get(2).trim + nrTailRegexp,
```
Comment (jo-pol) [inline src/main/scala/nl/knaw/dans/easy/bag2deposit/ddm/ReportRewriteRule.scala:45]: @janvanmansum 
You don't want column numbers but names. Our code base has only one [example](https://github.com/DANS-KNAW/easy-springfield/blob/104c1511047bdc09b8c2669b83a1c934fbca47ca/src/main/scala/nl.knaw.dans.easy.springfield/CreateSpringfieldActions.scala#L41-L43). To be consistent I would als have to apply it to the other CSV files. That seems to conflict with the provenance in their headers:
https://github.com/DANS-KNAW/easy-convert-bag-to-deposit/blob/0226a00019c349824af839541d41ac2c73de1d82/src/test/resources/debug-config/ABR-complex.csv#L1-L2
https://github.com/DANS-KNAW/easy-convert-bag-to-deposit/blob/0226a00019c349824af839541d41ac2c73de1d82/src/test/resources/debug-config/ABR-period.csv#L1-L2

Is it worth the trouble?

**Pull request merged**